services/application.js

```diff
@@ -97,25 +98,20 @@ class Application {
     async saveApplication(params, context) {
         verifySession(context)
             .verifyInitialized();
-        let application = params.application;
-        application.updatedAt = getCurrentTime();
-        const id = application?._id;
-        if (!id) return await this.createApplication(application, context.userInfo);
-        const aApplication = await this.getApplicationById(id);
-        const option = aApplication && aApplication.status !== IN_PROGRESS ? {$push: { history: HistoryEventBuilder.createEvent(context.userInfo._id, IN_PROGRESS, null)}}: null;
-        const result = await this.applicationCollection.update({...application, status: IN_PROGRESS}, option);
-        if (result.matchedCount < 1) throw new Error(ERROR.APPLICATION_NOT_FOUND+id);
-
-        const promises = [
-            // store application state change
-            await this.logCollection.insert(
-                UpdateApplicationStateEvent.create(context.userInfo._id, context.userInfo.email, context.userInfo.IDP, aApplication._id, aApplication.status, IN_PROGRESS)
-            )
-        ];
-        return await Promise.all(promises).then(function(results) {
-            return results[0]
-        });
-
+        let inputApplication = params.application;
+        inputApplication.updatedAt = getCurrentTime();
+        const id = inputApplication?._id;
+        if (!id) {
+            return await this.createApplication(inputApplication, context.userInfo);
+        }
+        const storedApplication = await this.getApplicationById(id);
+        const prevStatus = storedApplication?.status;
+        let application = {...storedApplication, ...inputApplication, status: IN_PROGRESS};
+        application = await updateApplication(this.applicationCollection, application, prevStatus, context?.userInfo?._id);
+        if (prevStatus !== application.status){
+            await logStateChange(this.logCollection, context.userInfo, application, prevStatus);
```

Comment:
I'm okay with using all asynchronous methods, but running four promises in a row might slow things down.

Reply:
In this instance it is going to run the 1 createApplication async method or the other 3 async methods. I think these need to be made synchronous with await because each relies on the success and return values or return values the previous call. If getApplicationByID does not find an application then an error should be thrown and we should not do the update operation. If the updateApplication fails then we do not want to log that an update took place